src/block/block.service.ts

```diff
@@ -0,0 +1,133 @@
+import { Injectable } from "@nestjs/common";
+import { providers } from "ethers";
+import { Block } from "src/models/Block";
+import { OperationFactory } from "src/models/OperationFactory";
+import { PartialBlockIdentifier } from "src/models/PartialBlockIdentifier";
+import { Transaction } from "src/models/Transaction";
+import { TransactionIdentifier } from "src/models/TransactionIdentifer";
+import { getRPCProvider } from "src/utils/client";
+
+import { RewardService } from "./reward.service";
+
+@Injectable()
+export class BlockService {
+  constructor(private rewardService: RewardService) {}
+
+  public async getBlock(block?: number) {
+    const provider = getRPCProvider();
+    const {
+      transactions,
+      miner,
+      timestamp,
+      parentHash,
+      hash,
+      number: blockNumber,
+    } = await provider.getBlockWithTransactions(block || "latest");
+
+    const parent = await provider.getBlock(parentHash);
+
+    const receipts = await Promise.all(
+      transactions.map((tx) => provider.getTransactionReceipt(tx.hash))
+    );
+
+    const operationFactory = new OperationFactory();
+
+    const blockTransaction = await this.buildBlockTransactions(
+      miner,
+      receipts,
+      transactions,
+      operationFactory
+    );
+
+    const rewardTransaction = await this.getRewardTransaction(
+      blockNumber,
+      miner,
+      operationFactory
+    );
+
+    return {
+      block: new Block(
+        new PartialBlockIdentifier(blockNumber, hash),
+        new PartialBlockIdentifier(parent.number, parent.hash),
+        timestamp,
+        [...blockTransaction, rewardTransaction]
+      ),
+    };
+  }
+
+  public async getBlockTransaction(
+    blockNumber: number,
+    transactionHash: string
+  ) {
+    const provider = getRPCProvider();
+
+    const transaction = await provider.getTransaction(transactionHash);
+
+    if (transaction.blockNumber !== blockNumber) {
+      throw new Error("Blocknumber mismatch");
+    }
+
+    const receipt = await transaction.wait();
+    const block = await provider.getBlock(blockNumber);
+
+    const blockTransaction = await this.buildBlockTransactions(
```

Comment:
if the value for the transaction is `0` it will be filtered out by the `buildBlockTransactions` function. Should we not return `0` value transactions when they are requested by hash?

Reply:
Yes. Good point, I think they should be returned as well.